{"cells":[{"cell_type":"markdown","metadata":{"id":"7d-SeF-yab93"},"source":["<img src=\"materials/images/introduction-to-proteomics-cover.png\"/>"]},{"cell_type":"markdown","metadata":{"id":"GJbxv7aG6RFa"},"source":["# Introduction to Proteomics"]},{"cell_type":"markdown","metadata":{"id":"fv630D1R6RFb"},"source":["`üïí This module should take less than 1 hour to complete.`\n","\n","`‚úçÔ∏è This notebook is written using Python.`"]},{"cell_type":"markdown","metadata":{"id":"TVc-fNRP9nom"},"source":["## What is protein?\n","\n","Proteins are the main product of our body, and present in all living organisms. Proteins are made up of amino acids. They are molecules that carry out the majority of biological functions, such as the formation of structural fibers of muscle tissue, enzymatic digestion of food, and synthesis and replication of DNA.\n","\n","There are two major types of proteins, **enzymes** and **protein hormones**:\n","\n","1. **Enzymes** accelerate chemical reactions. They help us to break down large molecules into smaller molecules.\n","\n","> For example, lactase is an enzyme protein that is produced in the small intestine. It is responsible for breaking down sugar molecule lactose into two products: glucose and fructose. When small intestines cannot produce enough lactase to break down the lactose, people often struggle with symptoms of lactose intolerance.\n","\n"]},{"cell_type":"markdown","metadata":{"id":"QdQfXEzhaq-z"},"source":["<img src=\"materials/images/enzyme-function.png\"/>"]},{"cell_type":"markdown","metadata":{"id":"MOT0NUvg-f4P"},"source":["2. **Protein hormones** help regulate the cell signaling processes so that cells in our physical bodies can maintain healthy conditions to carry out functions properly (i.e. homeostasis).\n","\n","> For example, when the blood sugar level goes up, pancreas produces a protein called insulin. What insulin does is that it signals fat cells to start storing sugar molecules (i.e. glucose molecules) in the blood to keep the amount of sugar molecules in the blood within a normal range. This will help to keep the glucose in the blood within a normal range. Let‚Äôs review the mechanism by using diabetes as an example. The symptoms are results of too much sugar building up in the blood, such as heavy thirst, blurry vision, and sugar in the urine. You may ask how the sugars end up accumulating in the blood?\n","\n"]},{"cell_type":"markdown","metadata":{"id":"bgNNKGlQb4kv"},"source":["<img src=\"materials/images/how-insulin-works.png\"/>"]},{"cell_type":"markdown","metadata":{"id":"vpkc0wyi_SBd"},"source":["In a healthy person‚Äôs cell, once insulin attaches to the insulin receptor on a cell, the cell opens its glucose channel to let glucose molecules in. The cells of a diabetic person fails to respond to insulin‚Äôs signal that asks the cell to open its glucose channel. This leaves all the glucose modules that are supposed to enter the cell accumulating in the blood. This is why scientist often refer to diabetic population as ‚Äúinsulin resistant population‚Äù in scientific publications - the cells in people with diabetes are not responding to insulin properly [1]."]},{"cell_type":"markdown","metadata":{"id":"tuY-B7xaazcQ"},"source":["<img src=\"materials/images/insulin-sensitive-vs-resistant.png\"/>"]},{"cell_type":"markdown","metadata":{"id":"qE9esIkGFqnN"},"source":["\n","\n","---\n","\n"]},{"cell_type":"markdown","metadata":{"id":"IdiJrRTyFEBA"},"source":["## What is a proteome?\n","\n","Proteome is a set of proteins produced in an organism. The term ‚Äúproteome‚Äù was first created by Australian scientist, Mark Wilkins, in 1994 to describe the set of proteins encoded by the genome, ‚ÄúPROTein complement expressed by a genOME‚Äù  [1][2].\n","\n","## What is proteomics?\n","\n","Proteomics is the large-scale study of proteins. Here are two examples of how to distinguish the use of proteome and proteomics:\n","\n","1. We could say Marc Wilkins is an expert in the field of proteomics.\n","2. Alternatively, we could say Marc Wilkins has published several research articles studying the proteome of human tissue samples.\n","\n"]},{"cell_type":"markdown","metadata":{"id":"BIdTmDhEFxj1"},"source":["\n","\n","---\n"]},{"cell_type":"markdown","metadata":{"id":"_ycSuYagF05y"},"source":["## How do we quantify protein or small molecules?\n","\n","In order to say whether someone‚Äôs blood sugar is high or low, we will need to quantify the proteins. There are a variety of protein quantitation methods (e.g., relative quantitation, absolute quantitation). Each of these protein quantitation technologies has unique benefits. An important method for **quantifying mass** and **characterizing proteins** is mass spectrometry.\n","\n","Mass spectrometry is an analytical tool useful for measuring the mass-to-charge ratio (m/z) of one or more molecules presented in a sample [3]. Each molecule has a given mass (m) and the charge number of ions (z). The m/z ratio is important information that helps us understand the identity of the molecule.\n"]},{"cell_type":"markdown","metadata":{"id":"QG_t3Wad9dd-"},"source":["---\n","\n"]},{"cell_type":"markdown","metadata":{"id":"QBzCPEMBF8N1"},"source":["## What are the major applications of proteomics?\n","\n","In personalized medicine, applications of quantifying proteomics are typically based on biomarkers, which are measurable indicators that reflect disease risk or disease status [4]. For example, studies show the proteomic profile of prostate cancer can be significantly altered during the course of the disease [5]."]},{"cell_type":"markdown","metadata":{"id":"qH210mDTa9yj"},"source":["<img src=\"materials/images/applications-of-proteomics.png\"/>"]},{"cell_type":"markdown","metadata":{"id":"QWhiChQtKE_w"},"source":["\n","\n","---\n","\n","\n"]},{"cell_type":"markdown","metadata":{"id":"zTRrnbJtOppy"},"source":["#### References:\n","\n","[1] Tyers, M., & Mann, M. (2003). From genomics to proteomics. Nature, 422(6928), 193-197.\n","\n","\n","[2] Marc Wilkins (geneticist). (2022). Retrieved August 25, 2022, from https://en.wikipedia.org/wiki/Marc_Wilkins_(geneticist)\n","\n","[3] https://www.broadinstitute.org/technology-areas/what-mass-spectrometry - last access: August 22, 2022.\n","\n","[4] Schubert, O. T., R√∂st, H. L., Collins, B. C., Rosenberger, G., & Aebersold, R. (2017). Quantitative proteomics: challenges and opportunities in basic and applied research. Nature protocols, 12(7), 1289-1294.\n","\n","[5] Latonen, L., Afyounian, E., Jylh√§, A., N√§ttinen, J., Aapola, U., Annala, M., ... & Visakorpi, T. (2018). Integrative proteomics in prostate cancer uncovers robustness against genomic and transcriptomic aberrations during disease progression. Nature communications, 9(1), 1-13.\n","\n","[6] https://www.genecards.org/cgi-bin/carddisp.pl?gene=CFD - last access: October 19, 2022.\n","\n"]},{"cell_type":"markdown","metadata":{"id":"msAl0H71RGp_"},"source":["---"]},{"cell_type":"markdown","metadata":{"id":"9UJGX28O9s9R"},"source":["# Proteomics Data Analysis\n","\n","### Background\n","\n","We will be focusing our analyses on a patient (i.e. Patient Z) from the Integrated Personal Omics Profiling (iPOP) study. Designed and performed at Stanford University, iPOP aims to understand what ‚Äúhealthy\" biochemical and physiological profiles look like at a personal level, and what happens when people become ill. This provides the basis for personalized precision medicine, which attempts to tailor medical decisions and treatments to a patient's individual omics profile.\n","\n","IPOP is a longitudinal study that follows 106 patients, and collects samples at regular intervals over several years through sickness and health. The iPOP cohort also contains a significant proportion of pre-diabetic patients. Therefore, another focus of the study was to better understand how omics are influenced by a pre-diabetic state and the progression from pre-diabetes to either a normal healthy state or diabetes.\n","\n","In this module, we will be reading and visualizing the proteomics data using Python. In our analyses, we will focus on a specific time period when Patient Z experienced an infection (and eventually recovered). By the end of this module, we will be able to interpret visualizations of the data and find interesting correlations."]},{"cell_type":"markdown","metadata":{"id":"MekiL1esJPL_"},"source":["### Load data\n","\n","Let's read our proteome abundances file named `abundance.csv` using the Python Pandas library. `index_col=0` sets the first entry in each row of the file as the row name.\n","\n","In our data, labeled row indices are the timepoint when each of the samples were collected (corresponding to different stages of the infection). Unlabeled (integer) row indices are other samples that were collected from Patient Z. Each column name represents a different protein."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"4qDc4kGG3N3t"},"outputs":[],"source":["import pandas as pd\n","\n","data = pd.read_csv('data/abundance.csv', index_col=0)\n","data"]},{"cell_type":"markdown","metadata":{"id":"VEiCysWgy7ed"},"source":["To get an overview of our data, we could take a look at the number of samples and proteins by examining the shape of the DataFrame."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"3BdeH3NKK7cV"},"outputs":[],"source":["(nRows, nCols) = data.shape\n","print('Our dataset contains the abundance data for', nCols, 'proteins from', nRows, 'samples.')"]},{"cell_type":"markdown","metadata":{"id":"010Qu-vM1Wu2"},"source":["### Checking our data\n","\n","Now, let's make sure our data makes sense. You do not have to worry to much about this code, but we are using a `for` loop and the `iloc` function to locate each row. `np.argmax` gets the index of the maximum value within each row. We can then print the most abundant protein and sample name by using our indices to access the indices (timepoints) and columns (protein names) from our data table."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"2XkXYKn8A2Q6"},"outputs":[],"source":["import numpy as np\n","\n","for i in range(5):\n","  j = np.argmax(data.iloc[i, :])\n","  print(data.columns[j], 'is the protein with the highest abundance in the', data.index[i], 'sample.')"]},{"cell_type":"markdown","metadata":{"id":"F_J_lgmKN2ov"},"source":["**Sanity Check**: ALB (albumin) is the most abundant protein in human blood. It checks out that each of our 5 labeled samples has albumin as the most abundant protein."]},{"cell_type":"markdown","metadata":{"id":"FO-UQeeCSqFY"},"source":["### Plotting\n","\n","Since the quantities of each of the proteins assayed vary in scale, we will scale these measurements to a value between 0 and 1. We obtain this relative abundance for each protein by subtracting the minimum abundance from all measurements the measurements of the specific protein. This gives us the minimum value of 0. Then, we can ensure the maximum value is 1 by dividing all of the shifted measurements by the maximum value (maximum abundance - minimum abundance)."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"xLUYo1xo0wRB"},"outputs":[],"source":["relativeAbundance = []\n","for protein in range(data.shape[1]):\n","  minAbundance = min(data.iloc[:, protein])\n","  maxAbundance = max(data.iloc[:, protein])\n","  relativeAbundance.append((data.iloc[:, protein] - minAbundance) / (maxAbundance - minAbundance))\n","relativeAbundance = pd.DataFrame(relativeAbundance)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"rwKXhHIjQtLO"},"outputs":[],"source":["relativeAbundance.iloc[0:10, :5]"]},{"cell_type":"markdown","metadata":{"id":"JKJ83RkNOs4H"},"source":["Now, we are ready to plot our data using a heatmap. Using the `seaborn` package, we can plot our relative abundances for the first 20 of our proteins of interest from our samples. We choose to plot this data using a heatmap, because it allows us to easily visualize changes in protein abundance over time (by simply comparing colors). Here, we choose to plot just the first 20 proteins to produce a reasonably sized heatmap. However, we could make additional plots for all the remaining proteins."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"XowYaZZQT37x"},"outputs":[],"source":["import seaborn as sns\n","\n","df = relativeAbundance.iloc[0:20, :5]\n","ax = sns.heatmap(df, vmin=0, vmax=1)\n","ax.figure"]},{"cell_type":"markdown","metadata":{"id":"7N1zkamHSgBm"},"source":["It is very difficult to draw conclusions from this heatmap, but we can observe that there is a visibly noticeable change in the heatmap profile when the infection recovery first starts. Preceding this, we see a low amount (dark cell) of the CFD (Complement Factor D) protein. CFD deficiency has been associated with recurrent bacterial meningitis infections for human patients [6]. We can rationalize this as the CFD levels increased back to earlier levels as the recovery from the infection progressed."]},{"cell_type":"markdown","metadata":{"id":"86mzH4DOOe3b"},"source":["\n","\n","---\n","\n"]},{"cell_type":"markdown","metadata":{"id":"tfF4unq1RJc0"},"source":["## Contributions & acknowledgement\n","\n","- **Module Content:** Ryan Park\n","- **Engineering:** Amit Dixit\n","- **UX/UI Design & Illustration:** Kexin Cha\n","- **Video Production:** Francesca Goncalves\n","- **Project Management:** Amir Bahmani, Kexin Cha"]},{"cell_type":"markdown","metadata":{"id":"oEqZp-EfQ8R1"},"source":["\n","\n","---\n","\n"]},{"cell_type":"markdown","metadata":{"id":"YsynKXR_SJCi"},"source":["Copyright (c) 2022 Stanford Data Ocean (SDO)\n","\n","All rights reserved."]}],"metadata":{"colab":{"provenance":[{"file_id":"1Gc7QzF7Ky3mMLlVEb1uDzXzQj9_LdJ0r","timestamp":1665169571198}]},"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.9"}},"nbformat":4,"nbformat_minor":0}